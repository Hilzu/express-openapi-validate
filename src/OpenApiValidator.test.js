"use strict";

const OpenApiValidator = require("./OpenApiValidator");

const minimumValidOpenApiDocument = {
  openapi: "3.0.1",
  info: { title: "Example API", version: "1.0.0" },
  paths: {},
};

const openApiDocument = Object.assign({}, minimumValidOpenApiDocument, {
  paths: {
    "/echo": {
      post: {
        description: "Echo input back",
        requestBody: {
          content: {
            "application/json": {
              schema: {
                type: "object",
                properties: {
                  input: {
                    type: "string",
                  },
                },
                required: ["input"],
              },
            },
          },
        },
        responses: {
          200: {
            content: {
              "application/json": {
                schema: {
                  type: "object",
                  properties: {
                    output: {
                      type: "string",
                    },
                  },
                  required: ["output"],
                },
              },
            },
          },
        },
      },
    },
  },
});

describe("OpenApiValidator", () => {
  test("can be created with valid OpenAPI document", () => {
    const validator = new OpenApiValidator(minimumValidOpenApiDocument);
    expect(validator).toBeInstanceOf(OpenApiValidator);
  });

  test("creating throws with Swagger 2.0 document", () => {
    expect(() => {
      // eslint-disable-next-line no-unused-vars
      const validator = new OpenApiValidator({
        swagger: "2.0",
        info: { title: "Swagger API", version: "1.0.0" },
        paths: {},
      });
    }).toThrowErrorMatchingSnapshot();
  });

  test("getting an operation object fails with invalid arguments", () => {
    const validator = new OpenApiValidator(openApiDocument);

    expect(() => {
      // eslint-disable-next-line no-unused-vars
      const op = validator.getOperationObject("POST", "/echo");
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      // eslint-disable-next-line no-unused-vars
      const op = validator.getOperationObject("ppost", "/echo");
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      // eslint-disable-next-line no-unused-vars
      const op = validator.getOperationObject("post", "/echoo");
    }).toThrowErrorMatchingSnapshot();
  });

  test("getting an operation object succeeds", () => {
    const validator = new OpenApiValidator(openApiDocument);
    const op = validator.getOperationObject("post", "/echo");
    expect(op).toEqual(openApiDocument.paths["/echo"].post);
  });

  test("validate", done => {
    const validator = new OpenApiValidator(openApiDocument);
    const validate = validator.validate("post", "/echo");
    validate({ body: { input: "hello" }}, {}, err => {
      expect(err).toBeUndefined();
      done();
    });
  });

  test("validate with invalid body", done => {
    const validator = new OpenApiValidator(openApiDocument);
    const validate = validator.validate("post", "/echo");
    validate({ body: {} }, {}, err => {
      expect(err).toBeInstanceOf(Error);
      expect(err).toMatchSnapshot();
      done();
    });
  });
});
