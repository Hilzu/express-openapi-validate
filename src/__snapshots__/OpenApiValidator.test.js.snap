// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpenApiValidator creating throws with Swagger 2.0 document 1`] = `"Unsupported OpenAPI / Swagger version=2.0"`;

exports[`OpenApiValidator getting an operation object fails with invalid arguments 1`] = `"Not a valid method=POST"`;

exports[`OpenApiValidator getting an operation object fails with invalid arguments 2`] = `"Not a valid method=ppost"`;

exports[`OpenApiValidator getting an operation object fails with invalid arguments 3`] = `"Path=/echoo with method=post not found from OpenAPI document"`;

exports[`OpenApiValidator validate with invalid body 1`] = `[ValidationError: Error while validating request parameters]`;
